### ISSUE TITLE
+[NSString stringWithFormat:]

### ISSUE THREAD

**Issue by charlieMonroe** [drkameleon/iswift#68]
This is pretty much as the [NSURL URLWithString:] issue I posted earlier, though in this case, it would be super cool if you could parse out the arguments and inline them into the format string.

E.g.

```
NSString *myVar = ...;
NSString *formattedString = [NSString stringWithFormat:@"Hello %i %@", 2, myVar];
```

would become

```
let myVar = ...
let formattedString = "Hello \(2) \(myVar)"
```

The same could apply to `-[NSString stringByAppendingFormat:]`...

**Comment by drkameleon** [comment]
Oh... good one. The truth is I have thought of that roughly 100 times, but all keeping postponing it.

The core issue (inlining arguments), has been posted and resolved already: https://github.com/drkameleon/iswift/issues/12

Now, the thing is:

It can obviously be done for `stringWithFormat:` and `stringByAppendingFormat:`.

But, what about a full list of methods (`NSString`-related or not) that could make use of the same thing? I mean, if I do it, it wouldn't hurt to track down all methods that make use of the same pattern, though I'm not sure I can find them all...

So, the question is: do I just start with `NSString`?

Any ideas are more than welcome! :)

**Comment by charlieMonroe** [comment]
Apple uses annotations for the format methods/functions. E.g.:

```
- (instancetype)initWithFormat:(NSString *)format, ... NS_FORMAT_FUNCTION(1,2);
```

Which means that format itself is the first argument, and the arglist starts with the second argument.

```
- (instancetype)initWithFormat:(NSString *)format arguments:(va_list)argList NS_FORMAT_FUNCTION(1,0);
```
The format is the first argument and takes no list of arguments (since it's wrapped in `va_list`).

```
- (instancetype)initWithFormat:(NSString *)format locale:(nullable id)locale, ... NS_FORMAT_FUNCTION(1,3);
```

Format is the first argument, argument list starts with the third argument.

I don't know if you are examining dynamically the header files of Foundation/Cocoa/UIKit, but this is how to find those methods/functions...

**Comment by drkameleon** [comment]
> I don't know if you are examining dynamically the header files of Foundation/Cocoa/UIKit, but this is how to find those methods/functions

Well, not exactly. But very close. For instance, almost all of the Cocoa-related types have been retrieved from the original documentation itself.

So, using that as a general pattern of getting this type of methods is not a bad idea at all - and why not, the header files themselves. Nice idea - I'll have a look into it...

**Comment by drkameleon** [comment]
Issue fixed as of the upcoming 1.9.

*For now, apart from `NSLog` which was already being handled, `stringWithFormat:` and `stringByAppendingFormat:` have also been added. More to follow!*

----

**Example:**

```objective-c
@implementation aClass

- (void)m {
    NSString* username = @"DrKameleon";
    NSString* formattedString = [NSString stringWithFormat:@"Hello %@!", username]; 
    NSString* appendFormat = [formattedString stringByAppendingFormat:@" - %@", [username uppercaseString]];
}

@end
```

**Output:**

```swift
class aClass {

	func m() {
		var username: String = "DrKameleon"
		var formattedString: String = "Hello \(username)!"
		var appendFormat: String = formattedString + " - \(username.uppercaseString())"
	
	}
	
}
```
